README.md
# domanek_project_2

## Aufgabenstellung

Netzwerk‐basiertes RobotGame wie robowiki mit zentralem Server.

## Spiel

Das Spiel ist eine abgeänderte Version von [robowiki](https://robowiki.net/wiki/Main_Page).

Die Ressourcen für das Spiel wurden von [robocode](https://github.com/robo-code/robocode) verwendet.

Das Spiel beginnt indem man den Game-Serevr startet:

`./robo-server`

Danach startet man in einem neuen Terminal jeweils einen weiteren Client.

1. `./rect-client`
2. `./random-client`
3. `./domanek`
4. `./domanekV2`

In dem CLient kann man dem Roboter folgende Befehle schicken  
- nach vorne fahren / nach hinten fahren / stoppen
- nach links drehen / nach rechts drehen / nicht mehr drehen
- Geschützturm nach links drehen / nach rechts drehen / nicht mehr drehen
- schießen / nicht mehr schießen
- man kann die eigene Position, Rotation und Geschützturmrotation erhalten
- man kann die ID, Name, Position, Rotation und Geschützturmrotation aller Gegnererhalten

Nachdem der Server gestartet wurde, können alle Clients beitreten.
Der Client legt seinen Namen und seine Farbe fest und schickt diese an den Server.
Sobald die maximale Anzahl der Spieler (-m,--max-players) erreicht wurde startet das Spiel.
Alle Clients bekommen ihre ID und die Roboter-Konfiguration.
Diese Konfiguration enhält die Geschwindigkeit, die Lebenspunkte, die Geschützturm- / Roboterrotationsgeschwindikeit, 
in welchem Abstand man schießen kann und ob man während dem Fahren schießen kann.
Das Spiel startet auch gleichzeitig und der RPC-Server wurde erstellt nun kann man die oben erwähten Befehle an den Server schicken.

Der Server startet nicht nur die grafische Oberfläche sondern zeigt auch im Terminal an, wenn Spielen beitreten und wenn Roboter sterben und am Ende wer gewonnen hat.

## einfaches Testen
Für den einfachsten Tests ob alles funktioniert benötigt man 3 Terminals in dem man jeweils einen dieser Befehle ausführt:

1. `./robo-server -m 2`
2. `./test-client`
3. `./test-client`

## Verwendung

Usage: ./robo-server [OPTIONS]

| Option                   | Typ [Bereich]=Standartwert    | Beschreibung                                 |
| ------------------------ | ----------------------------- | -------------------------------------------- |
| -h,--help                | FLAG                          | Print this help message and exit             |
| -w,--width               | UINT:INT in [500 - 1500]=950  | width and height of the window               |
| --bullet-speed           | FLOAT:INT in [1 - 50]=6       | the speed of the bullets                     |
| --bullet-damage          | FLOAT=20                      | the damage of the bullets                    |
| --bullet-size            | UINT:INT in [1 - 10]=3        | the size of the bullets in pixel             |
| --robot-speed            | FLOAT:FLOAT in [0.5 - 10]=2   | the speed of the robots                      |
| --health                 | FLOAT:POSITIVE=100            | the health of the robots                     |
| --robot-rotation         | FLOAT:FLOAT in [0.5 - 3.5]=1  | the roation speed of the robots              |
| --turret-rotation        | FLOAT:FLOAT in [0.5 - 5]=1.5  | the turret roation speed of the robots       |
| -f,--fire-countdown      | INT:POSITIVE=500              | the time after a robot can shoot again in ms |
| -m,--max-players         | UINT:INT in [2 - 4]=4         | the maximum players of the game              |
| -p,--port                | UINT=1113                     | port to connect to                           |
| -s,--not-shoot-and-move  | FLAG                          | Robot can not shoot while it is moving       |
| -j,--json-config         | TEXT:FILE                     | JSON Configuration for port, max-player, ... |

## verwendete Software
* [asio](https://think-async.com/Asio/)
* [CLI11](https://github.com/CLIUtils/CLI11)
* [fmt](https://github.com/fmtlib/fmt)
* [json](https://github.com/nlohmann/json)
* [spdlog](https://github.com/gabime/spdlog)

* [protocol-buffers](https://developers.google.com/protocol-buffers/)
* [gRPC](http://www.grpc.io/)

* [sfml](https://github.com/SFML/SFML)


CHANGELOG.org
* Changelog

** [1.0.0] - 2021-04-11
*** Added
- Added all robot functions to grpc server / client
- Added grpcServer/Client cpp file
- Added 4 example clients
- Added json to config
- Added documentation (Ausarbeitung)
*** Changed
- Changed directory structure
*** Removed
- 


** [0.1.1] - 2021-03-29
*** Added
- Added grpc
- Added grpcServer
- Added grpcClient
- Added grpcServer start thread
- Added robotClient 
*** Changed
-
*** Removed
- removed robot2 and robot 3 from main


** [0.1.0] - 2021-03-08
*** Added
- Added name of robot above robot (Text)
- Added health of robot below robot (Text)
- Added deathpoint of robot
- Added arial font
- Added RobotProperties to store individual properties about the robot
- Added RobotConfiguration to store global configuration about all robots
- Added max players to CLI and to game
- Added min fire countdown to CLI and game
*** Changed
- Changed all config values so that they must be passed to the constructor
- Changed body of robot to grey so client can choose color
*** Removed
- Removed all unnecessary values from the robot class and moved them in the config/properties class


** [0.0.4] - 2021-03-01
*** Added
- Added border for robot so that it does not drive outside the window
- Added collision between robots so that they can not collide
- Added robot with random movement
- Added robot that changes direction on wall hit
- Added health for robot
- Robots can die and a robot can win a game
- Adde
*** Changed
- Changed Window class to singelton
- Fixed error when bullet gets erased
*** Removed
- Removed second keyboard controlled robot


** [0.0.3] - 2021-02-22
*** Added
- Added robowiki resources
- Added keyboard controll for example application
- Added Robot class to control the robot
- Added Bullet class to shoot
- Added Window class to controll the graphical interface
*** Changed
- Changed Robot class to be more performant (cache movement vector)
- Changed Bullet class to be more performant (save movement in constructor)
*** Removed
- removed unnecessary template files


** [0.0.2] - 2021-02-15
*** Added
- Added smfl libary
- Added smfl example in main
*** Changed
- Changed Readme (Software)
- Changed meson build and options with new software
*** Removed
- Tested hello_imgui but removed it
- Tested tabulate but removed it


** [0.0.1] - 2021-02-08
*** Added
- Create repository and send invitation to prof
- Fill .gitignore
- Create empty doc folder
- Instantiate given template

server/src/main.cpp
#include "bullet.h"
#include "window.h"
#include "robot.h"
#include "robotConfiguration.h"
#include "robotProperties.h"
#include "bulletConfiguration.h"
#include "base64.h"

#include "grpcServer.h"
#include <grpcpp/health_check_service_interface.h>

#include "asio.hpp"
#include "CLI11.hpp"
#include <SFML/Graphics.hpp>
#include "spdlog/spdlog.h"
#include "spdlog/sinks/rotating_file_sink.h"

#include <vector>
#include <math.h>
#include <random>
#include <thread>


using namespace std;
using namespace asio;
using namespace asio::ip;

int main(int argc, char* argv[]) {
    GOOGLE_PROTOBUF_VERIFY_VERSION;

    CLI::App app("Robotgame");

    unsigned int width{950};
    app.add_option("-w,--width", width, "width and height of the window", true)->check(CLI::Range(500, 1500));

    float bulletSpeed{6};
    app.add_option("--bullet-speed", bulletSpeed, "the speed of the bullets", true)->check(CLI::Range(1, 50));

    float bulletDamage{20};
    app.add_option("--bullet-damage", bulletDamage, "the damage of the bullets", true);

    unsigned int bulletSize{3};
    app.add_option("--bullet-size", bulletSize, "the size of the bullets in px", true)->check(CLI::Range(1, 10));

    float robotSpeed{2};
    app.add_option("--robot-speed", robotSpeed, "the speed of the robots", true)->check(CLI::Range(0.5f, 10.0f));

    float health{100};
    app.add_option("--health", health, "the health of the robots", true)->check(CLI::PositiveNumber);

    float robotRotation{1};
    app.add_option("--robot-rotation", robotRotation, "the robot roation speed of the robots", true)->check(CLI::Range(0.5f, 3.5f));

    float turretRotation{1.5};
    app.add_option("--turret-rotation", turretRotation, "the turret roation speed of the robots", true)->check(CLI::Range(0.5f, 5.0f));

    int minFireCountdown{500};
    app.add_option("-f,--fire-countdown", minFireCountdown, "the time after a robot can shoot again in ms", true)->check(CLI::PositiveNumber);

    unsigned int maxPlayers{4};
    app.add_option("-m,--max-players", maxPlayers, "the maximum players of the game", true)->check(CLI::Range(2, 4));

    unsigned short  port{1113};
    app.add_option("-p,--port", port, "port to connect to", true);

    CLI11_PARSE(app, argc, argv);

    auto file_logger = spdlog::rotating_logger_mt("file_logger", "../logs/server.log", 1048576 * 5, 3);
    spdlog::set_default_logger(file_logger);
    spdlog::set_pattern("[%Y %m %d %H:%M:%S,%e] [%l] %v");
    spdlog::set_level(spdlog::level::debug);

    RobotConfiguration config{robotSpeed, health, robotRotation, turretRotation, minFireCountdown};
    BulletConfiguration bulletConfig{bulletSpeed, bulletDamage, bulletSize};

    vector<RobotProperties> properties;
    vector<tcp::iostream*> streams;

    for(unsigned int i = 0; i < maxPlayers; i++){
        try {
            io_context ctx;
            tcp::endpoint ep{tcp::v4(), port};
            tcp::acceptor acceptor{ctx, ep};
            
            acceptor.listen();
        
            tcp::iostream* strm{new tcp::iostream(acceptor.accept())};
            string data;
            *strm >> data;
            *strm << i << "\n";
            

            RobotPropertiesMessage rpmsg;
            rpmsg.ParseFromString(Base64::from_base64(data));

            streams.push_back(strm);

            RobotProperties property{rpmsg.name(), sf::Color(rpmsg.color())};
            properties.push_back(property);
            
            
        } catch (asio::system_error& e) {
            return 0;
        }
    }

    Window& window{Window::getInstance(width, maxPlayers, bulletConfig)};

    vector<Robot*> robots;

    for(RobotProperties property: properties) {
        Robot* robo = new Robot(property, config);
        robots.push_back(robo);
        window.addRobot(robo);
    }

    spdlog::info("Game started");
    fmt::print("Game started");

    std::thread t1{[&robots]{
        Robot_RPC_Server service{robots};
        grpc::EnableDefaultHealthCheckService(true);
        grpc::ServerBuilder builder;
        builder.AddListeningPort("0.0.0.0:50051", grpc::InsecureServerCredentials());
        builder.RegisterService(&service);
        std::unique_ptr<grpc::Server> server(builder.BuildAndStart());
        server->Wait();
    }};

    for(const auto strm: streams) {
        (*strm) << "start\n";
    }

    while (window.isOpen()) {
        window.clear();

        for(Robot* robo: robots) {
            if (!robo->isDead()) {
                robo->performActions();
            }
        }

        window.moveAllBullets();
        window.bulletHit();

        window.handleEvent();
        window.draw();     
        window.display();
    }

    t1.join();
}
